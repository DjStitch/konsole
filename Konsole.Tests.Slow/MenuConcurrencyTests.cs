using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using Konsole.Drawing;
using Konsole.Menus;
using NUnit.Framework;

namespace Konsole.Tests.Slow
{
    [UseReporter(typeof(DiffReporter))]
    public class MenuConcurrencyTests
    {
        private List<Task> _tasks = new List<Task>();

        [Test]
        public void SeperateThreadsForMenuAndTwoWindows()
        {
            var console = new MockConsole();
            var client = Window.Open(35, 0, 40, 25, "client", LineThickNess.Single, ConsoleColor.White, ConsoleColor.DarkBlue, console);
            var server = Window.Open(77, 0, 40, 25, "server", LineThickNess.Single, ConsoleColor.White, ConsoleColor.DarkYellow, console);
            // print two lines before the menu
            console.WriteLine("line 1");
            console.WriteLine("line 2");
            // create and run a menu inline, at the current cursor position
            var menu = new Menu(console,console, "ZeroMQ samples", ConsoleKey.Escape, 30,
                new MenuItem('1', "cats", () => RunMenuItem(client, "client", "cats")),
                new MenuItem('2', "dogs", () => RunMenuItem(server, "server", "dogs")),
                new MenuItem('3', "item 1", () => {}),
                new MenuItem('4', "item 2", () => {}),
                new MenuItem('5', "item 3", () => {}),
                new MenuItem('6', "item 4", () => {}),
                new MenuItem('7', "item 5", () => {})
                
                );
            // line below should print after (below) the menu.
            console.WriteLine("line 3");
            // console should continue working and cursor should be set to below the menu.
            var kb = new MockKeyboard(0, GetKeys()); ;

            menu.Keyboard = kb;
            menu.Run();
            Task.WaitAll(_tasks.ToArray());
            Approvals.Verify(console.BufferWrittenString);

        }

        void RunMenuItem(IConsole console, string service, string prefix)
        {
            var t = Task.Run(() =>
            {
                Console.WriteLine("starting " + service);
                for (int i = 0; i < 8000; i++) console.WriteLine("{0} {1}", prefix, i.ToString());
            });
            _tasks.Add(t);
        }


        IEnumerable<ConsoleKey> GetKeys()
        {
            yield return ConsoleKey.Enter;
            yield return ConsoleKey.DownArrow;
            yield return ConsoleKey.Enter;
            yield return ConsoleKey.DownArrow;
            

            for (int i = 0; i < 1000; i++)
            {
                yield return ConsoleKey.DownArrow;
                yield return ConsoleKey.DownArrow;
                yield return ConsoleKey.DownArrow;
                yield return ConsoleKey.UpArrow;
                yield return ConsoleKey.UpArrow;
            }
            yield return ConsoleKey.Escape;
        }

    }
}
