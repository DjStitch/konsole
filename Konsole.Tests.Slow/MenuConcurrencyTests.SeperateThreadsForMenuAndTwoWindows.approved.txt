                                   ┌─────────────── client ───────────────┐  ┌─────────────── server ───────────────┐   
    ZeroMQ samples                 │cats 7977                             │  │dogs 7977                             │   
    --------------------------     │cats 7978                             │  │dogs 7978                             │   
    cats                           │cats 7979                             │  │dogs 7979                             │   
    dogs                           │cats 7980                             │  │dogs 7980                             │   
    item 1                         │cats 7981                             │  │dogs 7981                             │   
    item 2                         │cats 7982                             │  │dogs 7982                             │   
    item 3                         │cats 7983                             │  │dogs 7983                             │   
    item 4                         │cats 7984                             │  │dogs 7984                             │   
    item 5                         │cats 7985                             │  │dogs 7985                             │   
                                   │cats 7986                             │  │dogs 7986                             │   
                                   │cats 7987                             │  │dogs 7987                             │   
                                   │cats 7988                             │  │dogs 7988                             │   
                                   │cats 7989                             │  │dogs 7989                             │   
                                   │cats 7990                             │  │dogs 7990                             │   
                                   │cats 7991                             │  │dogs 7991                             │   
                                   │cats 7992                             │  │dogs 7992                             │   
                                   │cats 7993                             │  │dogs 7993                             │   
                                   │cats 7994                             │  │dogs 7994                             │   
                                   │cats 7995                             │  │dogs 7995                             │   
                                   │cats 7996                             │  │dogs 7996                             │   
                                   │cats 7997                             │  │dogs 7997                             │   
                                   │cats 7998                             │  │dogs 7998                             │   
                                   │cats 7999                             │  │dogs 7999                             │   
                                   └──────────────────────────────────────┘  └──────────────────────────────────────┘   